collectors/python.d.plugin/openldap/openldap.chart.py

```diff
@@ -0,0 +1,203 @@
+    ),
+    'initiated_operations': (
+        'cn=Operations,cn=Monitor', 'monitorOpInitiated',
+    ),
+    'referrals_sent': (
+        'cn=Referrals,cn=Statistics,cn=Monitor', 'monitorCounter',
+    ),
+    'entries_sent': (
+        'cn=Entries,cn=Statistics,cn=Monitor', 'monitorCounter',
+    ),
+    'bind_operations': (
+        'cn=Bind,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'unbind_operations': (
+        'cn=Unbind,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'add_operations': (
+        'cn=Add,cn=Operations,cn=Monitor', 'monitorOpInitiated',
+    ),
+    'delete_operations':  (
+        'cn=Delete,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'modify_operations': (
+        'cn=Modify,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'compare_operations': (
+        'cn=Compare,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'search_operations': (
+        'cn=Search,cn=Operations,cn=Monitor', 'monitorOpCompleted',
+    ),
+    'write_waiters': (
+        'cn=Write,cn=Waiters,cn=Monitor', 'monitorCounter',
+    ),
+    'read_waiters': (
+        'cn=Read,cn=Waiters,cn=Monitor', 'monitorCounter',
+    ),
+}
+
+class Service(SimpleService):
+    def __init__(self, configuration=None, name=None):
+        SimpleService.__init__(self, configuration=configuration, name=name)
+        self.order = ORDER
+        self.definitions = CHARTS
+
+        self.server = configuration.get('server')
+        self.port = configuration.get('port')
+        self.username = configuration.get('username')
+        self.password = configuration.get('password')
+
+        self.alive = False
+        self.conn = None
+
+
+    def disconnect(self):
+        if self.conn:
+            self.conn.unbind()
+            self.conn = None
+            self.alive = False
+
+
+    def connect(self):
+        try:
+            self.conn = ldap.initialize('ldap://%s:%s' % ( self.server, self.port))
+            self.conn.simple_bind(self.username, self.password)
+        except ldap.LDAPError as error:
+            self.error(error)
+            return False
+
+        self.alive = True
+        return True
+
+
+    def reconnect(self):
+        self.disconnect()
+        return self.connect()
+
+
+    def check(self):
+        if not HAS_LDAP:
+            self.error("'python-ldap' package is needed")
+            return None
+
+        return self.connect() and self.get_data()
+
+
+    def get_data(self):
+        if not self.alive and not self.reconnect():
+            return None
+
+        data = dict()
+        for key in SEARCH_LIST.keys():
+            bind = SEARCH_LIST[key][0]
+            attr = SEARCH_LIST[key][1]
+            try:
+                num = self.conn.search(bind, ldap.SCOPE_BASE, 'objectClass=*', [attr,])
+                result_type, result_data = self.conn.result(num, 1)
+            except ldap.LDAPError as error:
+                self.error(error)
```

Comment:
I am not a `ldap` specialist but i think the only possible reason for exception here is connection problem (service restarted, service anavailable for some reason if it's remote `ldap` instance). correct me if am wrong.

Then we need to change `alive` flag to `False` here and return `None`. module wil attempt to reconnect on next iteration.

Reply:
You are right. I am just starrting to understand the logic behind connection management by the module. Thanks

Reply:
changed with 
```
            try:
                num = self.conn.search(bind, ldap.SCOPE_BASE, 'objectClass=*', [attr, ])
                result_type, result_data = self.conn.result(num, 1)
            except ldap.LDAPError as error:
                self.alive = False
                return None
```

Reply:
Looks good to me